### PR TITLE
Added GCU activation function

### PR DESCRIPTION
**Pull request by tareknaser360**

### CODE REVIEW

**Comment by shubham1206agra** [comment]
Can you provide the source of this activation function for confirmation of corectness?

**Comment by tareknaser360** [comment]
> Can you provide the source of this activation function for confirmation of corectness?

Sure. Check out this [paper](https://arxiv.org/abs/2108.12943).

File: src/mlpack/methods/ann/activation_functions/gcu_function.hpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+   * Computes the gcu function.
+   *
+   * @param x Input data.
+   * @param y The resulting output activation.
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Fn(const InputVecType& x, OutputVecType& y)
+  {
+    y = x % arma::cos(x);
+  }
```
Comment (shrit) [inline src/mlpack/methods/ann/activation_functions/gcu_function.hpp:51]: would you explain what is the difference between these two?

Comment (shrit) [reply src/mlpack/methods/ann/activation_functions/gcu_function.hpp:51]: In the first one, there is a multiply operation, while in the second it is a reminder operator that is used. Why?

**Review by shrit — CHANGES_REQUESTED** [review]

File: src/mlpack/methods/ann/activation_functions/gcu_function.hpp (hunk 1 above)
Comment (tareknaser360) [reply src/mlpack/methods/ann/activation_functions/gcu_function.hpp:51]: % is the operator for element-wise multiplication of two objects in Armadillo.

Comment (tareknaser360) [reply src/mlpack/methods/ann/activation_functions/gcu_function.hpp:51]: > would you explain what is the difference between these two?

First function takes a single value (double) as input. Second one takes a matrix as input.

File: src/mlpack/methods/ann/activation_functions/gcu_function.hpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
+/**
+ * @file methods/ann/activation_functions/gcu_function.hpp
+ * @author Tarek Elsayed
+ *
+ * Definition and implementation of the Growing Cosine Unit function.
```
Comment (zoq) [inline src/mlpack/methods/ann/activation_functions/gcu_function.hpp:5]: Can you add a reference to the paper, checkout https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/activation_functions/mish_function.hpp#L8 checkoutCan you add a reference to the paper, checkout https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/activation_functions/mish_function.hpp#L8 for an example.

File: src/mlpack/tests/ann/activation_functions_test.cpp [hunk 3]
```diff
@@ -785,4 +785,25 @@ TEST_CASE("SILUFunctionTest", "[ActivationFunctionsTest]")
 
   CheckActivationCorrect<SILUFunction>(activationData, desiredActivation);
   CheckDerivativeCorrect<SILUFunction>(desiredActivation, desiredDerivate);
+}
+
+/**
+ * Basic test of the GCU(Growing Cosine Unit) Function
```
Comment (zoq) [inline src/mlpack/tests/ann/activation_functions_test.cpp:791]:
```suggestion
 * Basic test of the GCU (Growing Cosine Unit) function.

```

**Review by zoq — COMMENTED** [review]
Can you test the function on a the XOR problem as mentioned in the paper?

**Comment by tareknaser360** [comment]
Can someone review this?

File: src/mlpack/methods/ann/activation_functions/gcu_function.hpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+ * @author Tarek Elsayed
+ *
+ * Definition and implementation of the Growing Cosine Unit function.
+ *
+ * For more information, see the following paper.
+ *
+ * @code
+ * @misc{
+ *   authors = {Mathew Mithra Noel, Arunkumar L, Advait Trivedi, Praneet Dutta},
+ *   title = {Growing Cosine Unit: A Novel Oscillatory Activation Function That Can Speedup Training and Reduce Parameters in Convolutional Neural Networks},
```
Comment (zoq) [inline src/mlpack/methods/ann/activation_functions/gcu_function.hpp:12]: Can you make sure this fits into our 80 character line limit; see https://github.com/mlpack/mlpack/wiki/DesignGuidelines#line-length-and-wrapping for more details.

**Comment by zoq** [comment]
Can you train a simple neural network (MLP) with the activation function on the XOR problem?

File: src/mlpack/methods/ann/activation_functions/gcu_function.hpp (hunk 4 above)
Comment (tareknaser360) [reply src/mlpack/methods/ann/activation_functions/gcu_function.hpp:12]: Sure

**Comment by tareknaser360** [comment]
> Can you train a simple neural network (MLP) with the activation function on the XOR problem?

How can I do that? Should I train it in a separate notebook in the examples repository?

**Comment by zoq** [comment]
You can use https://github.com/mlpack/mlpack/blob/master/src/mlpack/tests/ann/feedforward_network_test.cpp#L767-L800 as a baseline, swap out the activation function, and create a simple toy dataset (XOR).

**Comment by zoq** [comment]
I can provide more detailed information if you need some help e.g. regarding creating the toy dataset.

**Comment by tareknaser360** [comment]
Thank you.
I think I understand now. I will try implementing it.

**Comment by tareknaser360** [comment]
```
/**
 * Test that the GCU activation function solves
 * the XOR problem.
 */
TEST_CASE("GCUActivationFunction", "[FeedForwardNetworkTest]")
{
  // Load the dataset.
  arma::mat trainData;
  if (!data::Load("thyroid_train.csv", trainData))
    FAIL("Cannot open thyroid_train.csv");

  arma::mat trainLabels = trainData.row(trainData.n_rows - 1);
  trainData.shed_row(trainData.n_rows - 1);
  trainLabels -= 1; // The labels should be between 0 and numClasses.

  arma::mat testData;
  testData << -1 << -1 << arma::endr << 1 << -1 << arma::endr << -1 << 1
  << arma::endr << 1 << 1;

  arma::mat testLabels;
  testData << -1 << arma::endr << 1 << arma::endr << 1 << arma::endr << -1;

  // A single neuron network with GCU activation function.
  FFN<NegativeLogLikelihood> model;
  model.Add<Linear>(1);
  model.Add<GCUFunction>();
  model.Add<LogSoftMax>();

  ens::StandardSGD opt(0.1, 1, 5);

  double objVal = model.Train(trainData, trainLabels, opt);

  TestNetwork<>(model, trainData, trainLabels, testData, testLabels, 10, 0.1);
}

```
This what I have right now.
There are 3 issues:

   1- I can't add GCU or any activation function as a layer to the network because model.add() expects a layer as an argument.
   2- I can't evaluate the predictions using a call to TestNetwork() because it uses RMSProp and We need to use SGD.
   3-  I need to use the sign() function to map the output to either 1 or -1, I can't seem to find it in the layers.

**Comment by tareknaser360** [comment]
This is the description of the problem solution as per the original paper.

> The XOR problem was solved by using a single neuron with oscillatory activation functions, mean square loss and simple Stochastic Gradient Descent (SGD). 
> The initial weight vector was initialized with uniform random numbers in the interval [0 , 1].
> After training the output of the neuron is mapped to the class label using the signum function.

**Comment by zoq** [comment]
In the example `trainData` should be the XOR problem, the test data is the same, but this is okay, since we just want to make sure the activation function is able to solve the problem.

**Comment by zoq** [comment]
Regarding the layer issue, add the activation function to https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/layer/base_layer.hpp and follow the other examples.

**Comment by zoq** [comment]
`RMSProp` should be fine, it uses a slightly different update strategy.

**Comment by zoq** [comment]
Not completely sure what you mean with the map/sign function.

**Comment by tareknaser360** [comment]
> Not completely sure what you mean with the map/sign function.

It's mapping values to either 1 or -1.

**Comment by tareknaser360** [comment]
I added the GCU function as a layer.
I also copied the `TestNetwork` function and tweaked it to use SGD update rule and to use the signum function. 

So now we have a 1 layer neural network with the gcu function and sgd update rule and signum function as the paper suggested. 
I can't identify the issue now as it is failing 2 tests out of the 4 xor tests.

**Comment by zoq** [comment]
keep open
